external_ml/seviri/predictCPHCOT.py

```diff
@@ -0,0 +1,377 @@
+                ir134,     # 7    
+                lsm,       # 8    
+                nir016,    # 9    
+                skt,       # 10   
+                vis006,    # 11  
+                vis008,    # 12  
+                ir062,     # 13   
+                ir073      # 14 
+                ]
+ 
+    # check if array dimensions are equal throughout all arrays
+    # if all dimensions are equal: set dimension constants for reshaping
+    xdims = []
+    ydims = []
+    for tmp in data_lst:
+        xdims.append(tmp.shape[0])
+        ydims.append(tmp.shape[1])
+    
+    if hf.all_same(xdims) and hf.all_same(ydims):
+        xdim = data_lst[0].shape[0]
+        ydim = data_lst[0].shape[1]
+    else:
+        msg = 'xdim or ydim differ between input arrays for neural network.'
+        #logging.error(msg)
+        raise Exception(RuntimeError, msg)
+    
+    #fill neural network input array with flattened data fields
+    idata = np.empty((xdim*ydim, len(data_lst)))
+    has_invalid_item = np.empty((xdim*ydim))
+    
+    for cnt, d in enumerate(data_lst):
+        idata[:, cnt] = d.ravel()
+          
+    # check for each pixel if any channels is invalid (1), else 0
+    has_invalid_item = np.any(np.where(idata < 0, 1, 0), axis=1)
+    
+    all_chs = np.array([ir039, ir087, ir108, ir120, 
+               ir134, nir016, vis006, vis008, 
+               ir062, ir073])
+    
+    # pixels with all channels invalid = 1, else 0
+    all_channels_invalid = np.all(np.where(all_chs < 0, 1, 0), axis=0)
+    # indices of pixels with all channels valid
+    all_channels_valid_indxs = np.nonzero(~all_channels_invalid.ravel())
+    # dictionary of invalid pixel masks
+    masks = {'hii': has_invalid_item.reshape((xdim, ydim)), 
+             'aci': all_channels_invalid,
+             'acvi': all_channels_valid_indxs[0]}
+
+    scaled_data = {'COT': networks['COT'].scale_input(idata),
+                   'CPH': networks['CPH'].scale_input(idata)}
+
+    # apply scaling to input array
+    return scaled_data, (xdim, ydim), masks
+    
+
+def _select_networks():
+    """ Read driver file and select correct network. """
+    # read driver file for SEVIRI neural network
+    # assume driver file is in same directory as this file
+    basepath = os.path.dirname(os.path.realpath(__file__))
+    drifile = 'nn_driver.txt'
+    ptf = os.path.join(basepath, drifile)
+    logging.info("SEVIRI ANN DRIVER: {}".format(ptf))
+
+    opts = readdriver.parse_nn_driver(ptf)
+    
+    networks = {'COT': neuralnet.NetworkCOT(opts),
+                'CPH': neuralnet.NetworkCPH(opts)}
+    return networks
+
+
+def _thresholding(prediction, variable, opts):
+    """ Determine binary array by applying thresholding. """
+    # read threshold from driver file content
+    threshold = opts['NN_{}_THRESHOLD'.format(variable)]
+    
+    # appply threshold   
+    if variable == 'COT':
+        bin = np.where(prediction > threshold, IS_CLOUD, IS_CLEAR)
```

Comment:
The name `bin` should not be used as it shadows a built-in python function.

Reply:
Changed with ec34493.